File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/AllReady.java
package org.emulinker.kaillera.controller.v086.protocol;

import com.google.auto.value.AutoValue;
import java.nio.ByteBuffer;
import org.emulinker.kaillera.controller.messaging.*;
import org.emulinker.util.EmuUtil;

@AutoValue
public abstract class AllReady extends V086Message {
  public static final byte ID = 0x15;

  private static final String DESC = "All Ready Signal";

  public static AutoValue_AllReady create(int messageNumber) throws MessageFormatException {
    V086Message.validateMessageNumber(messageNumber, DESC);
    return new AutoValue_AllReady(messageNumber, ID, DESC);
  }

  @Override
  public int getBodyLength() {
    return 1;
  }

  @Override
  public void writeBodyTo(ByteBuffer buffer) {
    buffer.put((byte) 0x00);
  }

  public static AllReady parse(int messageNumber, ByteBuffer buffer)
      throws ParseException, MessageFormatException {
    if (buffer.remaining() < 1) throw new ParseException("Failed byte count validation!");

    byte b = buffer.get();

    if (b != 0x00)
      throw new MessageFormatException(
          "Invalid " + DESC + " format: byte 0 = " + EmuUtil.byteToHex(b));

    return AllReady.create(messageNumber);
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/AllReady.kt
package org.emulinker.kaillera.controller.v086.protocol

import java.nio.ByteBuffer
import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.messaging.ParseException
import org.emulinker.util.EmuUtil

data class AllReady
    @Throws(MessageFormatException::class)
    constructor(override val messageNumber: Int) : V086Message() {
  override val bodyLength = 1

  override val description = DESC
  override val messageId = ID

  init {
    validateMessageNumber(messageNumber, DESC)
  }

  public override fun writeBodyTo(buffer: ByteBuffer) {
    buffer.put(0x00.toByte())
  }

  companion object {
    const val ID: Byte = 0x15
    private const val DESC = "All Ready Signal"

    @Throws(ParseException::class, MessageFormatException::class)
    fun parse(messageNumber: Int, buffer: ByteBuffer): AllReady {
      if (buffer.remaining() < 1) throw ParseException("Failed byte count validation!")
      val b = buffer.get()
      if (b.toInt() != 0x00)
          throw MessageFormatException(
              "Invalid " + DESC + " format: byte 0 = " + EmuUtil.byteToHex(b))
      return AllReady(messageNumber)
    }
  }
}
